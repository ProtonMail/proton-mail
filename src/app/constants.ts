import { c } from 'ttag';
import { MAILBOX_LABEL_IDS } from 'proton-shared/lib/constants';

export const PAGE_SIZE = 50;

export const ELEMENT_TYPES = {
    MESSAGE: 'message',
    CONVERSATION: 'conversation'
};

export const LABEL_IDS_TO_HUMAN = {
    [MAILBOX_LABEL_IDS.INBOX]: 'inbox',
    [MAILBOX_LABEL_IDS.ALL_DRAFTS]: 'all-drafts',
    [MAILBOX_LABEL_IDS.ALL_SENT]: 'all-sent',
    [MAILBOX_LABEL_IDS.TRASH]: 'trash',
    [MAILBOX_LABEL_IDS.SPAM]: 'spam',
    [MAILBOX_LABEL_IDS.ALL_MAIL]: 'all-mail',
    [MAILBOX_LABEL_IDS.ARCHIVE]: 'archive',
    [MAILBOX_LABEL_IDS.SENT]: 'sent',
    [MAILBOX_LABEL_IDS.DRAFTS]: 'drafts',
    [MAILBOX_LABEL_IDS.STARRED]: 'starred'
};

export const HUMAN_TO_LABEL_IDS = Object.entries(LABEL_IDS_TO_HUMAN).reduce((acc, [key, value]) => {
    acc[value] = key;
    return acc;
}, Object.create(null));

export const LABEL_IDS_TO_I18N = {
    [MAILBOX_LABEL_IDS.INBOX]: c('Link').t`Inbox`,
    [MAILBOX_LABEL_IDS.ALL_DRAFTS]: c('Link').t`Drafts`,
    [MAILBOX_LABEL_IDS.ALL_SENT]: c('Link').t`Sent`,
    [MAILBOX_LABEL_IDS.TRASH]: c('Link').t`Trash`,
    [MAILBOX_LABEL_IDS.SPAM]: c('Link').t`Spam`,
    [MAILBOX_LABEL_IDS.ALL_MAIL]: c('Link').t`All mail`,
    [MAILBOX_LABEL_IDS.ARCHIVE]: c('Link').t`Archive`,
    [MAILBOX_LABEL_IDS.SENT]: c('Link').t`Sent`,
    [MAILBOX_LABEL_IDS.DRAFTS]: c('Link').t`Drafts`,
    [MAILBOX_LABEL_IDS.STARRED]: c('Link').t`Starred`
};

export const SIGNATURE_START = 1546300800; // January 1, 2019

export const VERIFICATION_STATUS = {
    NOT_VERIFIED: -1,
    NOT_SIGNED: 0,
    SIGNED_AND_VALID: 1,
    SIGNED_AND_INVALID: 2,
    SIGNED_NO_PUB_KEY: 3
};

export const MESSAGE_FLAGS = {
    FLAG_RECEIVED: 1, // whether a message is received
    FLAG_SENT: 2, // whether a message is sent
    FLAG_INTERNAL: 4, // whether the message is between ProtonMail recipients
    FLAG_E2E: 8, // whether the message is end-to-end encrypted
    FLAG_AUTO: 16, // whether the message is an autoresponse
    FLAG_REPLIED: 32, // whether the message is replied to
    FLAG_REPLIEDALL: 64, // whether the message is replied all to
    FLAG_FORWARDED: 128, // whether the message is forwarded
    FLAG_AUTOREPLIED: 256, // whether the message has been responded to with an autoresponse
    FLAG_IMPORTED: 512, // whether the message is an import
    FLAG_OPENED: 1024, // whether the message has ever been opened by the user
    FLAG_RECEIPT_SENT: 2048, // whether a read receipt has been sent in response to the message
    // For drafts only
    FLAG_RECEIPT_REQUEST: 65536, // whether to request a read receipt for the message
    FLAG_PUBLIC_KEY: 131072, // whether to attach the public key
    FLAG_SIGN: 262144 // whether to sign the message
};

export enum ENCRYPTED_STATUS {
    PGP_MIME = 8 // Used for attachment
}

export const AES256 = 'aes256';

export enum MESSAGE_ACTIONS {
    NEW = -1,
    REPLY = 0,
    REPLY_ALL = 1,
    FORWARD = 2
}
