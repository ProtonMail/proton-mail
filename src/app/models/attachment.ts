import { Binary } from './utils';

export interface Attachment {
    ID?: string;
    Name?: string;
    Size?: number;
    Preview?: any;
    KeyPackets?: any;
    MIMEType?: string;
    data?: any;
    Headers?: { [key: string]: string };
    Encrypted?: number;
    Signature?: any;
}

export interface AttachmentMime {
    checksum?: string;
    content: Binary;
    contentDisposition?: string;
    contentId?: string;
    contentType?: string;
    fileName?: string;
    generatedFileName?: string;
    length?: number;
    transferEncoding?: string;
}
