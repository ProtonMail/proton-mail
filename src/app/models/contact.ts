export interface ContactOrGroup {
    contact?: ContactEmail;
    group?: ContactGroup;
}

export interface Contact {
    ID?: string;
    Name?: string;
    UID?: string;
    Size?: number;
    CreateTime?: number;
    ModifyTime?: number;
    ContactEmails?: ContactEmail[];
    LabelIDs?: string[];
    Cards?: Card[];
}

export interface ContactEmail {
    ID?: string;
    Name?: string;
    Email?: string;
    Type?: string;
    Defaults?: number;
    Order?: number;
    ContactID?: string;
    LabelIDs?: string[];
}

export interface ContactGroup {
    ID?: string;
    Name?: string;
    Color?: string;
    Path?: string;
    Display?: number;
    Exclusive?: number;
    Notify?: number;
    Order?: number;
    Type?: number;
}

export interface Card {
    Type?: number;
    Data?: string;
    Signature?: string;
}

export type ContactCache = Map<string, Contact>;

export type ContactEmailCache = Map<string, ContactEmail>;
