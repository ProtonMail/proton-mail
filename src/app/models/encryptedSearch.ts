import { Message } from 'proton-shared/lib/interfaces/mail/Message';
import { DecryptedKey } from 'proton-shared/lib/interfaces';
import { DBSchema } from 'idb';
import { SearchParameters } from './tools';
import { Element } from './element';

export interface ESMetricsReport {
    numMessagesIDB: number;
    sizeIDBOnDisk?: number;
    sizeIDB: number;
    sizeCache: number;
    numMessagesSearched: number;
    searchTime: number;
    numMessagesFound: number;
    isFirstSearch: boolean;
    isCacheLimited: boolean;
}

export type MessageForSearch = Pick<
    Message,
    | 'ID'
    | 'Order'
    | 'ConversationID'
    | 'Subject'
    | 'Unread'
    | 'Sender'
    | 'AddressID'
    | 'Flags'
    | 'IsReplied'
    | 'IsRepliedAll'
    | 'IsForwarded'
    | 'ToList'
    | 'CCList'
    | 'BCCList'
    | 'Time'
    | 'Size'
    | 'NumAttachments'
    | 'ExpirationTime'
    | 'LabelIDs'
>;

export interface CachedMessage extends MessageForSearch {
    decryptedBody?: string;
    decryptedSubject?: string;
    decryptionError: boolean;
}

export interface AesGcmCiphertext {
    iv: Uint8Array;
    ciphertext: ArrayBuffer;
}

export interface StoredCiphertext extends Pick<MessageForSearch, 'ID' | 'LabelIDs' | 'Time' | 'Order'> {
    aesGcmCiphertext: AesGcmCiphertext;
}

export interface EncryptedSearchDB extends DBSchema {
    messages: {
        key: string;
        value: StoredCiphertext;
        indexes: { byTime: number[] };
    };
}

export interface RecoveryPoint {
    ID: string;
    Time: number;
}

export interface LastEmail {
    Time: number;
    Order: number;
}

export interface NormalisedSearchParams extends Omit<SearchParameters, 'wildcard' | 'keyword'> {
    labelID: string;
    normalisedKeywords: string[] | undefined;
    decryptionError?: boolean;
}

export interface ESDBStatus {
    dbExists: boolean;
    isBuilding: boolean;
    isDBLimited: boolean;
    esEnabled: boolean;
    isCacheReady: boolean;
    isCacheLimited: boolean;
    isRefreshing: boolean;
}

export interface ESSearchStatus {
    permanentResults: MessageForSearch[];
    setElementsCache: (Elements: Element[]) => void;
    labelID: string;
    cachePromise: Promise<CachedMessage[]>;
}

export type GetUserKeys = () => Promise<DecryptedKey[]>;

export type EncryptedSearch = (
    searchParams: SearchParameters,
    labelID: string,
    setCache: (Elements: Element[]) => void
) => Promise<boolean>;

export type CacheIndexedDB = (force?: boolean) => Promise<{ cachedMessages: CachedMessage[]; isCacheLimited: boolean }>;

export interface EncryptedSearchFunctions {
    encryptedSearch: EncryptedSearch;
    cacheIndexedDB: CacheIndexedDB;
    getESDBStatus: () => ESDBStatus;
    getProgressRecorderRef: () => React.MutableRefObject<[number, number]>;
    toggleEncryptedSearch: () => void;
    resumeIndexing: () => Promise<void>;
    pauseIndexing: () => Promise<void>;
}
