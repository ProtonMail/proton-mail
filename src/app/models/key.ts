import { PmcryptoKey } from 'pmcrypto';

export interface KeyData {
    Code: number;
    RecipientType: number;
    MIMEType: string;
    Keys: Key[];
    SignedKeyList: any[];
    Warnings: any[];
}

export interface Key {
    ID: string;
    Primary: number;
    Flags: number;
    Fingerprint: string;
    Fingerprints: string[];
    PublicKey: string;
    Version: number;
    Activation: any;
    PrivateKey: string;
    Token: any;
    Signature: any;
}

export interface AddressKeys {
    Key: Key;
    privateKey: PmcryptoKey;
    publicKey: PmcryptoKey;
}
