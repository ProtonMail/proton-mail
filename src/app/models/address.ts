import { ContactGroup } from './contact';

export interface Address {
    ID?: string;
    Email?: string;
    DisplayName?: string;
    Priority?: number;
    Receive?: number;
    Status?: number;
    Order?: number;
    Send?: number;
    Signature?: string;
}

export type RecipientType = 'ToList' | 'CCList' | 'BCCList';
export const recipientTypes: RecipientType[] = ['ToList', 'CCList', 'BCCList'];

export interface Recipient {
    Name?: string;
    Address?: string;
    Group?: string;
}

export interface RecipientGroup {
    group?: ContactGroup;
    recipients: Recipient[];
}

export interface RecipientOrGroup {
    recipient?: Recipient;
    group?: RecipientGroup;
}
