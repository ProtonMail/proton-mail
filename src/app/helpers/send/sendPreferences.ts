import { getKeys, decodeBase64, binaryStringToArray, getFingerprint, isExpiredKey, PmcryptoKey } from 'pmcrypto';

import { unique } from 'proton-shared/lib/helpers/array';
import { PACKAGE_TYPE, RECIPIENT_TYPE, MIME_TYPES, KEY_FLAGS } from 'proton-shared/lib/constants';

import { Message } from '../../models/message';
import { isEO, isSign } from '../message/messages';
import { normalizeEmail } from '../addresses';
import { Address } from '../../models/address';
import { getByEmail, isOwnAddress, isFallbackAddress } from '../addresses';
import { KeyData, Key } from '../../models/key';
import { findEmailInCache } from '../contacts';
import { ContactEmailCache, ContactEmail } from '../../models/contact';
import { base64ToArray } from '../base64';

export type MapPreference = { [email: string]: SendPreference };

export interface SendPreference {
    encrypt: boolean;
    sign: boolean;
    mimetype: MIME_TYPES;
    publickeys: PmcryptoKey[];
    primaryPinned: boolean;
    scheme: PACKAGE_TYPE;
    pinned: boolean;
    isVerified: boolean;
    warnings: any[];
    ownAddress: boolean;
}

type CachePreference = { [contactID: string]: { [email: string]: PreferenceInfo } };

interface PreferenceInfo {
    encryptFlag: boolean;
    signFlag: boolean;
    emailKeys: string[];
    mimetype: MIME_TYPES;
    scheme: PACKAGE_TYPE;
    isVerified: boolean;
}

const cache: CachePreference = {};

const emailInExtrInfo = (contactEmail: ContactEmail) =>
    cache[contactEmail.ContactID || ''] && cache[contactEmail.ContactID || ''][normalizeEmail(contactEmail.Email)];

const getInExtrInfo = (contactEmail: ContactEmail) =>
    (cache[contactEmail.ContactID || ''] || {})[normalizeEmail(contactEmail.Email)] || {};

const usesDefaults = (contactEmail: ContactEmail) => !contactEmail || contactEmail.Defaults;

/**
 * Determines if a certain key object is allowed to be used for encryption
 */
const encryptionEnabled = ({ Flags }: Key) => Flags & KEY_FLAGS.ENABLE_ENCRYPTION;

/**
 * Returns the default send preferences if no contact is available for the specified email address.
 * The global settings, composer mode and API keys can still change the defaults though.
 */
const getDefaultInfo = async (
    email: string,
    { RecipientType, Keys = [], Warnings: warnings = [] }: KeyData,
    defaultMimeType: MIME_TYPES,
    eoEnabled: boolean,
    globalSign: boolean,
    mailSettings: any,
    addresses: Address[]
): Promise<SendPreference> => {
    const isInternal = RecipientType === RECIPIENT_TYPE.TYPE_INTERNAL;
    const isExternal = RecipientType === RECIPIENT_TYPE.TYPE_EXTERNAL;
    const settingsScheme = mailSettings.PGPScheme;
    const settingsMime = settingsScheme === PACKAGE_TYPE.SEND_PGP_MIME ? MIME_TYPES.MIME : MIME_TYPES.PLAINTEXT;
    const address = getByEmail(addresses, email);
    const ownAddress = isOwnAddress(address, Keys);

    if ((isInternal || isExternal) && Keys.length) {
        const fallbackAddress = isFallbackAddress(address, Keys);

        return {
            warnings,
            encrypt: true,
            sign: true,
            mimetype: isExternal ? settingsMime : defaultMimeType,
            publickeys: await getKeys(Keys[0].PublicKey),
            primaryPinned: !fallbackAddress,
            scheme: isInternal ? PACKAGE_TYPE.SEND_PM : settingsScheme,
            pinned: ownAddress,
            ownAddress,
            isVerified: true
        };
    }
    if (eoEnabled) {
        return {
            warnings,
            encrypt: true,
            sign: false,
            mimetype: defaultMimeType,
            publickeys: [],
            primaryPinned: true,
            scheme: PACKAGE_TYPE.SEND_EO,
            pinned: false,
            ownAddress,
            isVerified: true
        };
    }
    return {
        warnings,
        encrypt: false,
        sign: globalSign,
        mimetype: globalSign ? settingsMime : defaultMimeType,
        publickeys: [],
        primaryPinned: true,
        scheme: globalSign ? settingsScheme : PACKAGE_TYPE.SEND_CLEAR,
        pinned: false,
        ownAddress,
        isVerified: true
    };
};

const mimetypeLogic = (
    mimetype: MIME_TYPES,
    defaultMimetype: MIME_TYPES,
    scheme: PACKAGE_TYPE,
    sign: boolean,
    encrypt: boolean
) => {
    /*
     * PGP/MIME can only send using the MIME encoding as it doesn't support separate attachments and we need to encode
     * them in the body
     */
    if (scheme === PACKAGE_TYPE.SEND_PGP_MIME && (sign || encrypt)) {
        return MIME_TYPES.MIME;
    }
    if (scheme === PACKAGE_TYPE.SEND_PGP_INLINE && (sign || encrypt)) {
        return MIME_TYPES.PLAINTEXT;
    }
    // If sending EO, respect the mime type of the composer, since it will be what the API returns when retrieving the message.
    if (scheme === PACKAGE_TYPE.SEND_EO) {
        return defaultMimetype;
    }
    if (defaultMimetype === MIME_TYPES.PLAINTEXT || mimetype === null) {
        // NEVER upconvert
        return defaultMimetype;
    }
    return mimetype;
};

/**
 * Checks if one of the allowed sending keys is pinned. This function returns true if key pinning is disabled
 * or if atleast on of the Sending keys are in the contacts
 * Should be done on extract, so API changes (the other user resetting their password) are noticed.
 */
const isPrimaryPinned = async (base64Keys: string[], Keys: Key[], email: string, addresses: Address[]) => {
    if (base64Keys.length === 0) {
        const address = getByEmail(addresses, email);
        return !isFallbackAddress(address, Keys);
    }

    const sendKeys = Keys.filter(encryptionEnabled).map((key) => key.PublicKey);
    const keys = await Promise.all(sendKeys.map(getKeys));
    const sendKeyObjects = keys.filter(([k = false]: any) => !!k);
    const [pinnedKey] = await getKeys(base64ToArray(base64Keys[0]) as any);
    const pinnedFingerprint = getFingerprint(pinnedKey);

    return (
        sendKeyObjects.length === 0 || sendKeyObjects.map(([k]: any) => getFingerprint(k)).includes(pinnedFingerprint)
    );
};

/**
 * Generates the sendpreferences using the extracted information after parsing the contacts.
 * This function is the counterpart extractInfo
 */
const extractInfo = async (
    { encryptFlag, signFlag, mimetype: inputMimetype, emailKeys, scheme: inputScheme, isVerified }: PreferenceInfo,
    keyData: KeyData,
    defaultMimeType: MIME_TYPES,
    eoEnabled: boolean,
    globalSign: boolean,
    email: string,
    mailSettings: any,
    addresses: Address[]
): Promise<SendPreference> => {
    const { RecipientType, Warnings = [], Keys = [] } = keyData;
    const isInternal = RecipientType === RECIPIENT_TYPE.TYPE_INTERNAL;
    const isExternalWithKeys = RecipientType === RECIPIENT_TYPE.TYPE_EXTERNAL && Keys.length > 0;
    const primaryPinned =
        isInternal || isExternalWithKeys ? await isPrimaryPinned(emailKeys, Keys, email, addresses) : true;
    const pmKey = isInternal || isExternalWithKeys ? await getKeys(Keys[0].PublicKey) : [];
    // In case the pgp packet list contains multiple keys, only the first one is taken.
    const keyObjs = await Promise.all(
        emailKeys
            .map((s) => decodeBase64(s) || '')
            .map(binaryStringToArray)
            .map((a) => {
                return getKeys(a).then(([k]) => isExpiredKey(k).then((isExpired: boolean) => (isExpired ? null : [k])));
            })
    );
    const keyObjects = (keyObjs.filter((k) => k !== null) as unknown) as PmcryptoKey[];

    const publickeys = keyObjects.length && primaryPinned ? [keyObjects[0]] : pmKey;
    const warnings = Warnings;
    let encrypt = isInternal || isExternalWithKeys || (encryptFlag && !!keyObjects.length);
    let sign = isInternal || isExternalWithKeys || (signFlag === null ? !!globalSign : signFlag);
    sign = sign || encryptFlag;

    let scheme: PACKAGE_TYPE;
    if (isInternal) {
        scheme = PACKAGE_TYPE.SEND_PM;
    } else {
        scheme = sign || encrypt ? inputScheme : PACKAGE_TYPE.SEND_CLEAR;
    }
    scheme = scheme === null ? mailSettings.PGPScheme : scheme;

    if (eoEnabled && !encrypt) {
        sign = false;
        encrypt = true;
        scheme = PACKAGE_TYPE.SEND_EO;
    }

    const mimetype = mimetypeLogic(inputMimetype, defaultMimeType, scheme, sign, encrypt);
    const pinned = keyObjects.length > 0;
    const ownAddress = false;

    return { publickeys, warnings, encrypt, sign, scheme, mimetype, primaryPinned, isVerified, pinned, ownAddress };
};

/**
 * Extracts the preferences from the contacts and stores it in the cache for reusage.
 * The logic is straightforward but we see for more info
 * https://docs.google.com/document/d/1lEBkG0DC5FOWlumInKtu4a9Cc1Eszp48ZhFy9UpPQso
 */
const getApiInfo = async (
    email: string,
    keyData: KeyData,
    defaultMimeType: MIME_TYPES,
    eoEnabled: boolean,
    globalSign: boolean,
    mailSettings: any,
    addresses: Address[],
    cache: ContactEmailCache
) => {
    const normalizedEmail = normalizeEmail(email);
    // const isInternal = keyData.RecipientType === RECIPIENT_TYPE.TYPE_INTERNAL;
    // const isExternalWithKeys = keyData.RecipientType === RECIPIENT_TYPE.TYPE_EXTERNAL && keyData.Keys.length > 0;

    const contactEmail = findEmailInCache(cache, normalizedEmail);
    if (usesDefaults(contactEmail)) {
        const info = await getDefaultInfo(
            email,
            keyData,
            defaultMimeType,
            eoEnabled,
            globalSign,
            mailSettings,
            addresses
        );
        return { [email]: info };
    }

    // TODO: Decrypt contacts

    ////////////////////////////////////
    ////////////// MOCK ////////////////
    ////////////////////////////////////

    const info = await getDefaultInfo(email, keyData, defaultMimeType, eoEnabled, globalSign, mailSettings, addresses);
    return { [email]: info };

    ////////////////////////////////////
    ////////////// MOCK ////////////////
    ////////////////////////////////////

    // const { vCard, errors } = await Contact.get(contactEmail.ContactID);

    // const keyList = toList(vCard.get('key'));
    // const encryptFlagList = toList(vCard.get('x-pm-encrypt'));
    // const signFlagList = toList(vCard.get('x-pm-sign'));
    // const schemeList = toList(vCard.get('x-pm-scheme'));
    // const mimeList = toList(vCard.get('x-pm-mimetype'));
    // const emailList = toList(vCard.get('email'));

    // const group = getGroup(emailList, normalizedEmail);
    // if (!group) {
    //     return { [email]: await getDefaultInfo(email, keyData, defaultMimeType, eoEnabled, globalSign) };
    // }

    // const matchesGroup = groupMatcher(group.toLowerCase());
    // const emailKeys = _.filter(keyList, matchesGroup);
    // const encryptFlag = _.find(encryptFlagList, matchesGroup);
    // const signFlag = _.find(signFlagList, matchesGroup);
    // const mimetypeProp = _.find(mimeList, matchesGroup);
    // const mimetype = mimetypeProp ? mimetypeProp.valueOf() : null;
    // const schemeProp = _.find(schemeList, matchesGroup);
    // const scheme = schemeProp ? toSchemeConstant(schemeProp.valueOf()) : null;
    // const base64Keys = await reorderKeys(
    //     keyData,
    //     (await Promise.all(_.map(emailKeys, (prop) => contactKey.getBase64Value(prop)))).filter(Boolean) // In case the key is expired or revoked we don't get the base 64 value but false
    // );
    // const data = {
    //     encryptFlag:
    //         isInternal ||
    //         isExternalWithKeys ||
    //         ((encryptFlag ? encryptFlag.valueOf().toLowerCase() !== 'false' : false) && emailKeys.length > 0),
    //     signFlag:
    //         isInternal ||
    //         isExternalWithKeys ||
    //         (signFlag ? signFlag.valueOf().toLowerCase() !== 'false' : !!globalSign),
    //     emailKeys: base64Keys,
    //     mimetype: mimetype !== 'text/plain' && mimetype !== 'text/html' ? null : mimetype,
    //     scheme: isInternal ? PACKAGE_TYPE.SEND_PM : scheme,
    //     isVerified: !errors.includes(CONTACT_ERROR.TYPE2_CONTACT_VERIFICATION)
    // };

    // // We don't support encryption without signing
    // data.signFlag = data.signFlag || data.encryptFlag;

    // CACHE.EXTRACTED_INFO[contactEmail.ContactID] = CACHE[contactEmail.ContactID] || {};
    // CACHE.EXTRACTED_INFO[contactEmail.ContactID][normalizedEmail] = data;

    // return { [email]: await extractInfo(data, keyData, defaultMimeType, eoEnabled, globalSign, email) };
};

/**
 * Extracts send preferences from the cache if available
 */
const getCacheInfo = async (
    email: string,
    keyData: KeyData,
    defaultMimeType: MIME_TYPES,
    eoEnabled: boolean,
    globalSign: boolean,
    mailSettings: any,
    addresses: Address[],
    cache: ContactEmailCache
): Promise<MapPreference> => {
    const normalizedEmail = normalizeEmail(email);
    const contactEmail = findEmailInCache(cache, normalizedEmail);

    if (usesDefaults(contactEmail)) {
        return {
            [email]: await getDefaultInfo(
                email,
                keyData,
                defaultMimeType,
                eoEnabled,
                globalSign,
                mailSettings,
                addresses
            )
        };
    }

    if (!emailInExtrInfo(contactEmail)) {
        // return { [email]: null };
        return {};
    }

    return {
        [email]: await extractInfo(
            getInExtrInfo(contactEmail),
            keyData,
            defaultMimeType,
            eoEnabled,
            globalSign,
            email,
            mailSettings,
            addresses
        )
    };
};

const inExtractedInfoCache = (contactEmailList: ContactEmail[]) =>
    contactEmailList.every((e) => usesDefaults(e) || emailInExtrInfo(e));

const inCache = (emails: string[], cache: ContactEmailCache) => {
    const normalizedEmails = emails.map(normalizeEmail);
    const contactEmailList = normalizedEmails.map((email) => findEmailInCache(cache, email));

    return inExtractedInfoCache(contactEmailList);
};

const getInfo = async (
    email: string,
    keyData: KeyData,
    defaultMimeType: MIME_TYPES,
    eoEnabled: boolean,
    globalSign: boolean,
    mailSettings: any,
    addresses: Address[],
    cache: ContactEmailCache
): Promise<MapPreference> => {
    const address = getByEmail(addresses, email);

    if (isOwnAddress(address, keyData.Keys)) {
        const info = await getDefaultInfo(
            email,
            keyData,
            defaultMimeType,
            eoEnabled,
            globalSign,
            mailSettings,
            addresses
        );
        return { [email]: info };
    }

    if (inCache([email], cache)) {
        return getCacheInfo(email, keyData, defaultMimeType, eoEnabled, globalSign, mailSettings, addresses, cache);
    }

    return getApiInfo(email, keyData, defaultMimeType, eoEnabled, globalSign, mailSettings, addresses, cache);
};

/**
 * The goal of this service is to provide all the encryption + encoding preferences for a recipient by parsing the
 * contact of the recipient, considering the general settings, inputs from the message that we want to send and API stuff
 *
 * For the general logic see:
 * https://docs.google.com/document/d/1lEBkG0DC5FOWlumInKtu4a9Cc1Eszp48ZhFy9UpPQso
 * This is the specification it should implement and should be the right way to do this
 *
 * primaryPinned basically just says if the primary key is available for sending (so either pinned or key pinning is disabled
 * It differs from pinned as pinned just says is key pinning is enabled.
 * primaryPinned is a flag that tells the FE that we first need to fix the sendPreference before sending.
 */
export const getSendPreferences = async (
    emails: string[] = [],
    message: Message,
    mailSettings: any,
    addresses: Address[],
    cache: ContactEmailCache,
    getPublicKeys: (email: string) => Promise<KeyData>,
    catchErrors = false
): Promise<MapPreference> => {
    const defaultMimeType: MIME_TYPES = message ? (message.MIMEType as MIME_TYPES) : MIME_TYPES.DEFAULT;
    const eoEnabled = isEO(message);
    const globalSign = message ? isSign(message) : mailSettings.Sign;
    const normEmails: string[] = unique(emails.map(normalizeEmail));
    const normInfos = await Promise.all(
        normEmails.map(async (email) => {
            try {
                const keyData = await getPublicKeys(email);
                return getInfo(email, keyData, defaultMimeType, eoEnabled, globalSign, mailSettings, addresses, cache);
            } catch (e) {
                if (!catchErrors) {
                    throw e;
                }
            }
        })
    );
    const normMap = Object.assign({}, ...normInfos);

    return emails.reduce((acc, cur) => {
        const result = normMap[normalizeEmail(cur)];
        if (!result) {
            return acc;
        }
        acc[cur] = result;
        return acc;
    }, {} as MapPreference);
};
