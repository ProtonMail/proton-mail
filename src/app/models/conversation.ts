import { Label } from './label';
import { Recipient } from './address';

export interface Conversation {
    ID?: string;
    Subject?: string;
    Size?: number;
    Time?: number;
    ContextTime?: number;
    NumMessages?: number;
    Senders?: Recipient[];
    Recipients?: Recipient[];
    NumUnread?: number;
    Labels?: Label[];
    LabelIDs?: string[];
    Order?: number;
}
