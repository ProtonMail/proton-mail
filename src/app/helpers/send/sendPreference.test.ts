import { getSendPreferences } from './sendPreferences';
import { RECIPIENT_TYPE } from 'proton-shared/lib/constants';
import { Address } from '../../models/address';
import { ContactEmail } from '../../models/contact';
import { KeyData, Key } from '../../models/key';

const ownEmail = 'own@test.com';
const email1 = 'test1@test.com';
const email2 = 'test2@test.com';
const email3 = 'test3@test.com';
const i = 4;
const emailCacheBuster = () => `test${i}@test.com`;
const message = {};
const mailSettings = {};
const ownAddress = { ID: 'own', Email: ownEmail, DisplayName: 'Own', Receive: 1 };
const addresses: Address[] = [ownAddress];
const contactCache = new Map<string, ContactEmail>();
const getPublicKeys = jest.fn(async () => ({} as KeyData));

describe('sendPreference', () => {
    afterEach(() => {
        jest.clearAllMocks();
    });

    const setup = async ({ emails = [emailCacheBuster()] } = {}) => {
        return getSendPreferences(emails, message, mailSettings, addresses, contactCache, getPublicKeys, false);
    };

    describe('result structure', () => {
        it('should return an empty object if no email', async () => {
            const result = await setup({ emails: [] });
            expect(Object.values(result).length).toBe(0);
        });

        it('should return one preference if one email', async () => {
            const result = await setup({ emails: [email1] });
            expect(Object.values(result).length).toBe(1);
            expect(result[email1]).toBeDefined();
        });

        it('should return as many preferences as emails count', async () => {
            const emails = [email1, email2, email3];
            const result = await setup({ emails });
            expect(Object.values(result).length).toBe(3);
            emails.forEach((email) => expect(result[email]).toBeDefined());
        });
    });

    describe('default', () => {
        it('should match default preferences', async () => {
            const result = await setup({ emails: [email1] });
            expect(result).toMatchSnapshot();
        });
    });

    describe('encrypt', () => {
        it('should not be encrypted by default', async () => {
            const result = await setup();
            expect(Object.values(result)[0].encrypt).toBe(false);
        });

        it('should be encrypted if internal', async () => {
            getPublicKeys.mockImplementation(
                async () =>
                    ({
                        RecipientType: RECIPIENT_TYPE.TYPE_INTERNAL,
                        Keys: [{}] as Key[]
                    } as KeyData)
            );

            const result = await setup();
            expect(Object.values(result)[0].encrypt).toBe(true);
        });

        it('should be encrypted if external', async () => {
            getPublicKeys.mockImplementation(
                async () =>
                    ({
                        RecipientType: RECIPIENT_TYPE.TYPE_EXTERNAL,
                        Keys: [{}] as Key[]
                    } as KeyData)
            );

            const result = await setup();
            expect(Object.values(result)[0].encrypt).toBe(true);
        });
    });

    describe('ownAddress', () => {
        it('should not be ownAddress by default', async () => {
            const result = await setup();
            expect(Object.values(result)[0].ownAddress).toBe(false);
        });

        it('should detect if its sent to a own address', async () => {
            const result = await setup({ emails: [ownEmail] });
            expect(Object.values(result)[0].ownAddress).toBe(true);
        });
    });
});
