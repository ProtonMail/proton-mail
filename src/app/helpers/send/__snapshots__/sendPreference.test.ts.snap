// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendPreference default should match default preferences 1`] = `
Object {
  "test1@test.com": Object {
    "encrypt": false,
    "isVerified": true,
    "mimetype": undefined,
    "ownAddress": false,
    "pinned": false,
    "primaryPinned": true,
    "publickeys": Array [],
    "scheme": 4,
    "sign": false,
    "warnings": Array [],
  },
}
`;
